clang/lib/Sema/SemaDecl.cpp

```diff
@@ -6888,6 +6888,42 @@ static void SetNestedNameSpecifier(Sema &S, DeclaratorDecl *DD, Declarator &D) {
   DD->setQualifierInfo(SS.getWithLocInContext(S.Context));
 }
 
+void Sema::deduceHLSLAddressSpace(VarDecl *Decl) {
+  // The variable already has an address space (groupshared for ex).
+  if (Decl->getType().hasAddressSpace())
+    return;
+
+  if (Decl->getType()->isDependentType())
+    return;
+
+  QualType Type = Decl->getType();
+  if (Type->isSamplerT() || Type->isVoidType())
+    return;
+
+  // Template instantiations can lack definition. In such case,
+  // we cannot deduce the AS.
+  // FIXME: figure out why RWBuffer<float> yields such declaration.
+  if (const RecordType *RT =
+          dyn_cast<RecordType>(Type->getUnqualifiedDesugaredType())) {
+    CXXRecordDecl *RD = Type->getAsCXXRecordDecl();
+    if (RD && !RD->isCompleteDefinition())
+      return;
+  }
+
+  // Resource handles.
+  if (Type->isHLSLIntangibleType())
```

Comment:
```suggestion
  if (isResourceRecordTypeOrArrayOf(Type->getUnqualifiedDesugaredType()))
```
`isResourceRecordTypeOrArrayOf` is a static function in SemaHLSL.cpp.